_topics/en/op_codeseparator.md

```diff
@@ -0,0 +1,150 @@
+  people who can spend it---without Alice giving away her private key or
+  creating an onchain transaction.  Bob can then delegate further by
+  adding Carol to the set of allowed spenders.  No released version of
+  Bitcoin supported this capability but source code claimed to be from
+  a pre-release version of Bitcoin [may][pre todd1] have [supported][pre
+  todd2] this capability.
+
+  In [Bitcoin 0.3.7][] (July 2010), a fix to the [1 OP_RETURN][] bug was
+  deployed as a stealth [hard fork][lukejr says sipa confirms hf] that
+  changed how the scripts were evaluated, including discontinuing the
+  use of the virtual `OP_CODESEPARATOR`.  However, the opcode itself
+  wasn't changed and scripts have continued to be able to include it
+  directly.
+
+  ### Later changes and proposals
+
+  The [BIP143][] signature verification rules for the initial version of
+  segwit (v0) activated in August 2017 specified a slightly different
+  version of `OP_CODESEPARATOR`.  In legacy script, the
+  `FindAndDelete()` operation is used to delete all code separators from
+  the scriptCode.  In segwit v0, the `FindAndDelete()` operation is
+  no longer performed, so any `OP_CODESEPARATOR` opcodes that appear
+  later in script than the most recently executed code separator will be
+  included in the scriptCode.  Otherwise, the opcode operates the same
+  by deleting its calling `OP_CODESEPARATOR` and any preceding parts of
+  the script from the scriptCode.
+
+  As of this writing, the proposed [tapscript][topic tapscript] soft
+  fork to implement v1 segwit script contains a further modified
+  `OP_CODESEPARATOR`.  In tapscript, signatures no longer commit to a
+  computed scriptCode; instead they indirectly commit to the executed
+  script through its leaf in the [taproot][topic taproot] merkle tree
+  (called the tapleaf).  The would render `OP_CODESEPARATOR` useless, so
+  in tapscript signatures must instead commit to the position of the
+  most recently executed `OP_CODESEPARATOR` (or 0xffffffff if none has
+  been executed).
+
+  Also as of this writing, the proposed [consensus cleanup][topic
+  consensus cleanup] soft fork specifies disabling the ability to use
+  `OP_CODESEPARATOR` in legacy Bitcoin scripts.  This would prevent it
+  from being used to call the relatively slow `FindAndDelete()`
+  operation an excessive amount of times, a problem known to result in
+  transactions and blocks that could take a long time to verify.  This
+  proposed change would make unspendable any money spent to a script
+  using code separator, so this proposed change met with some
+  [criticism][o'connor spendability] even though no one is known to be
+  using legacy `OP_CODESEPARATOR`.  It's unclear whether or not this
+  part of the proposal may be changed.
+
+  Additionally, Bitcoin Core 0.16.1 and later will not relay or mine any
+  transaction that uses `OP_CODESEPARATOR` in a legacy script.
```

Comment:
Great section. I'm trying to figure where commit https://github.com/bitcoin/bitcoin/commit/9dabfe49c066301ef75bcfcb089fd308366127c4 in PR https://github.com/bitcoin/bitcoin/pull/11423 fits into all this.
```markdown
 Add constant scriptCode policy in non-segwit scripts

This disables OP_CODESEPARATOR in non-segwit scripts (even in an unexecuted branch), and makes
a positive FindAndDelete result invalid. This ensures that the scriptCode serialized in
SignatureHash() is always the same as the script passing to the EvalScript.
```

Reply:
I'm not sure I understand your comment, but jl2012's commit basically makes the change proposed in the consensus cleanup section of the topic post and for the same reason, although obviously Lau's commit is only a standardness change and not consensus.

Reply:
Thanks!